### PR TITLE
add gspo

### PR DESCRIPTION
**Pull request by samsja**
https://www.arxiv.org/abs/2507.18071
screenshot-20250726-131342

### CODE REVIEW

File: src/prime_rl/trainer/config.py [hunk 1]
```diff
@@ -119,7 +119,13 @@ class RatioConfig(BaseGRPOVariantConfig):
...
 
 
-GRPOVariantsConfig: TypeAlias = Union[ClippingConfig, RatioConfig]
+class GSPOConfig(BaseGRPOVariantConfig):
+    type: Literal["gspo"] = "gspo"
+    epsilon: float = 3e-4
+    clip_ratio: Annotated[float, Field(ge=0)] = 8.0  # TODO delete but fur some reason does not pass
+
+
+GRPOVariantsConfig: TypeAlias = Union[ClippingConfig, RatioConfig, GSPOConfig]
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/config.py:128]: it doesnt pass because the "union" config must be a discrimnated union not just a union type alias for pydantic to resolve this properly. you can see how to do it properly in the buffer type config on orchestratror

File: src/prime_rl/trainer/loss.py [hunk 2]
```diff
@@ -42,6 +42,16 @@ def grpo_loss(
             clip_ratio=grpo_loss_config.clip_ratio,
             highest_entropy_percentage=grpo_loss_config.highest_entropy_ratio_loss,
         )
+    elif isinstance(grpo_loss_config, GSPOConfig):
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/loss.py:45]: uff, should renamem the discrimnated union to just `LossConfig`, otherwise this reads confusing

File: src/prime_rl/trainer/loss.py [hunk 3]
```diff
@@ -126,6 +136,41 @@ def grpo_loss_ratio(
...
+    epsilon: float,
+) -> tuple[Tensor, Tensor, Tensor]:
+    # Divide logits by sampling temperature.
+    # See https://huggingface.co/blog/the_n_implementation_details_of_rlhf_with_ppo#policy-training-implementation-details
+    shifted_logits = shifted_logits / temperature
+    per_token_logps = selective_log_softmax(shifted_logits, input_ids)
+
+    sequence_level_log_likelihood = (per_token_logps * loss_mask).sum(-1)
+    original_sequence_level_log_likelihood = (original_logprobs * loss_mask).sum(-1)
+    seq_len = loss_mask.sum(-1).clamp(min=1)
```
Comment (mikasenghaas) [inline src/prime_rl/trainer/loss.py:156]: when would we need to min clamp to 1 here?

**Review by mikasenghaas — COMMENTED** [review]
looks nice implementation-wise, super interested to see how it will go

**Comment by mikasenghaas** [comment]
closed in favor of #645
